File: RRT-Connect/Python/rrt_connect.py
"""
  @File: rrt-connect.py
  @Brief: RRT-Connect algorithm for pathplanning
 
  @Author: Benxiaogu
  @Github: https://github.com/Benxiaogu
  @CSDN: https://blog.csdn.net/weixin_51995147?type=blog
 
  @Date: 2024-11-13
"""
import numpy as np
import random
import math
from itertools import combinations
import matplotlib.pyplot as plt
import matplotlib.animation as animation
import matplotlib.patches as patches

class RRTConnect:
    def __init__(self,start,goal,obstacles,board_size,max_try,max_dist,goal_sample_rate,env) -> None:
        self.start = self.Node(start,None,0)
        self.goal = self.Node(goal,None,0)
        self.obstacles = obstacles
        self.board_size = board_size
        self.max_try = max_try # Number of iterations
        self.max_dist = max_dist # Maximum sampling distance
        self.goal_sample_rate = goal_sample_rate
        self.env = env
        self.inflation = 1
        self.searched = []

    class Node:
        def __init__(self,position,parent,cost) -> None:
            self.position = position
            self.parent = parent
            self.cost = cost


    def run(self):
        cost,path,expand = self.plan()
        self.searched = expand
        self.visualize(cost,path)

    def plan(self):
        nodes_forward = {self.start.position: self.start}
        nodes_back = {self.goal.position: self.goal}

        for iter in range(self.max_try):
            # Generate a random node
            node_rand = self.get_random_node()
            # Get the nearest neighbor node
            node_near = self.get_nearest_neighbor(list(nodes_forward.values()),node_rand)
            # Get the new node
            node_new = self.get_new_node(node_rand,node_near)
            
            if node_new:
                nodes_forward[node_new.position] = node_new
                node_near_b = self.get_nearest_neighbor(list(nodes_back.values()), node_new)
                node_new_b = self.get_new_node(node_new,node_near_b)

                if node_new_b:
                    nodes_back[node_new_b.position] = node_new_b
                    # Greedy extending
                    while True:
                        for node_position, node in nodes_back.items():
                            if node.position == node_new.position:
                                print("final")
                                cost, path = self.extractPath(node_new, nodes_back, nodes_forward)
                                expand = self.get_expand(list(nodes_back.values()), list(nodes_forward.values()))
                                print("Exploring {} nodes.".format(iter))
                                return cost, path, expand

                        node_new_b2 = self.get_new_node(node_new,node_new_b)

                        if node_new_b2:
                            nodes_back[node_new_b2.position] = node_new_b2
                            node_new_b = node_new_b2
                        else:
                            break

            if len(nodes_back) < len(nodes_forward):
                nodes_forward, nodes_back = nodes_back, nodes_forward

        return 0, None, None

    def get_random_node(self):
        """
            Return a random node.
        """
        if random.random()>self.goal_sample_rate:
            node = self.Node(
                (random.uniform(0,self.env.height),random.uniform(0,self.env.width)),None,0)
        else:
            node = self.goal
            
        return node

    def get_nearest_neighbor(self,node_list,node):
        """
            Return node that is nearest to 'node' in node_list
        """
        dist = [self.distance(node, n) for n in node_list]
        node_near = node_list[int(np.argmin(dist))]
        return node_near
    
    def get_new_node(self,node_rand,node_near):
        """
            Return node found based on node_near and node_rand.
        """
        dx = node_rand.position[0] - node_near.position[0]
        dy = node_rand.position[1] - node_near.position[1]
        dist = math.hypot(dx,dy)
        theta = math.atan2(dy, dx)

        d = min(self.max_dist,dist)
        position = ((node_near.position[0]+d*math.cos(theta)),node_near.position[1]+d*math.sin(theta))

        node_new = self.Node(position,node_near,node_near.cost+d)

        if self.isCollision(node_new, node_near):
            return None
        return node_new
    
    def isCollision(self,node1,node2):
        """
            Judge collision from node1 to node2 
        """
        if self.isInObstacles(node1) or self.isInObstacles(node2):
            return True
        
        for rect in self.env.obs_rectangle:
            if self.isInterRect(node1,node2,rect):
                return True
        
        for circle in self.env.obs_circle:
            if self.isInterCircle(node1,node2,circle):
                return True
        
        return False
    
    def distance(self,node1,node2):
        dx = node2.position[0] - node1.position[0]
        dy = node2.position[1] - node1.position[1]

        return math.hypot(dx,dy)
        
    def isInObstacles(self,node):
        """
            Determine whether it is in obstacles or not.
        """
        x,y = node.position[0],node.position[1]
        for (ox,oy,w,h) in self.env.boundary:
            if ox-self.inflation<x<ox+w+self.inflation and oy-self.inflation<y<oy+h+self.inflation:
                return True
        
        for (ox,oy,w,h) in self.env.obs_rectangle:
            if ox-self.inflation<x<ox+w+self.inflation and oy-self.inflation<y<oy+h+self.inflation:
                return True
            
        for (ox,oy,r) in self.env.obs_circle:
            if math.hypot(x-ox,y-oy)<=r+self.inflation:
                return True
        
        return False
    
    def isInterRect(self,node1,node2,rect):
        """"
            Judge whether it will cross the rectangle when moving from node1 to node2
        
        """
        ox,oy,w,h = rect
        vertex = [[ox-self.inflation,oy-self.inflation],
                  [ox+w+self.inflation,oy-self.inflation],
                  [ox+w+self.inflation,oy+h+self.inflation],
                  [ox-self.inflation,oy+h+self.inflation]]
        
        x1,y1 = node1.position
        x2,y2 = node2.position

        def cross(p1,p2,p3):
            x1 = p2[0]-p1[0]
            y1 = p2[1]-p1[1]
            x2 = p3[0]-p1[0]
            y2 = p3[1]-p1[0]
            return x1*y2 - x2*y1
        
        for v1,v2 in combinations(vertex,2):
            if max(x1,x2) >= min(v1[0],v2[0]) and min(x1,x2)<=max(v1[0],v2[0]) and \
                max(y1,y2) >= min(v1[1],v2[1]) and min(y1,y2) <= max(v1[1],v2[1]):
                if cross(v1,v2,node1.position) * cross(v1,v2,node2.position)<=0 and \
                    cross(node1.position,node2.position,v1) * cross(node1.position,node2.position,v2)<=0:
                    return True
    
        return False
    
    def isInterCircle(self,node1,node2,circle):
        """
            Judge whether it will cross the circle when moving from node1 to node2
        """
        ox,oy,r = circle

        dx = node2.position[0] - node1.position[0]
        dy = node2.position[1] - node1.position[1]

        # print("isInterCircle-dx:",dx)
        # print("isInterCircle-dy:",dy)
        d = dx * dx + dy * dy
        if d==0:
            return False
            
        # Projection
        t = ((ox - node1.position[0]) * dx + (oy - node1.position[1]) * dy) / d
        
        # The projection point is on line segment AB
        if 0 <= t <= 1:
            closest_x = node1.position[0] + t * dx
            closest_y = node1.position[1] + t * dy
        
            # Distance from center of the circle to line segment AB
            distance = math.hypot(ox-closest_x,oy-closest_y)
            
            return distance <= r+self.inflation
        
        return False

    def extractPath(self, node_middle, nodes_back, nodes_forward):
        """"
            Extract the path based on the closed set.
        """
        if self.start.position in nodes_back:
            nodes_forward, nodes_back = nodes_back, nodes_forward

        # forward
        node = nodes_forward[node_middle.position]
        path_forward = [node.position]
        cost = node.cost
        while node.position != self.start.position:
            node_parent = nodes_forward[node.parent.position]
            node = node_parent
            path_forward.append(node.position)

        # backward
        node = nodes_back[node_middle.position]
        path_back = []
        cost += node.cost
        while node.position != self.goal.position:
            node_parent = nodes_back[node.parent.position]
            node = node_parent
            path_back.append(node.position)
        
        path = list(reversed(path_forward))+path_back

        return cost, path
    
    def get_expand(self, nodes_back, nodes_forward):
        expand = []
        tree_size = max(len(nodes_forward), len(nodes_back))
        for tr in range(tree_size):
            if tr < len(nodes_forward):
                expand.append(nodes_forward[tr])
            if tr < len(nodes_back):
                expand.append(nodes_back[tr])

        return expand
    
    def visualize(self, cost, path):
        """
            Plot the map.
        """
        fig, ax = plt.subplots()
        ax.set_xlim(-self.board_size, self.env.width+self.board_size)
        ax.set_ylim(-self.board_size, self.env.height+self.board_size)

        plt.title("RRT\n"+"Cost: "+str(cost))

        # Draw rectangle obstacles
        for (ox, oy, w, h) in self.env.obs_rectangle:
            ax.add_patch(patches.Rectangle((ox, oy), w, h, edgecolor='black', facecolor='gray'))

        # Draw circle obstacles
        for (ox, oy, r) in self.env.obs_circle:
            ax.add_patch(patches.Circle((ox, oy), r, edgecolor='black', facecolor='gray'))

        # Draw boundary
        for (ox, oy, w, h) in self.env.boundary:
            ax.add_patch(patches.Rectangle((ox, oy), w, h, edgecolor='black', facecolor='black'))

        # Draw start anf goal
        ax.plot(self.start.position[0], self.start.position[1], 'go', markersize=8, label='Start')
        ax.plot(self.goal.position[0], self.goal.position[1], 'bo', markersize=8, label='Goal')

        path_line, = ax.plot([], [], c='red', linewidth=3,label='Path')

        # Draw the random tree dynamically
        if self.start.position in self.searched:
            self.searched.remove(self.start.position)
        if self.goal.position in self.searched:
            self.searched.remove(self.goal.position)
        def update(frame):
            if frame < len(self.searched):
                node = self.searched[frame]
                if node.parent is not None:
                    parent = node.parent
                    ax.plot([node.position[0], parent.position[0]], [node.position[1], parent.position[1]], 'g-', alpha=0.3)
            elif path:
                path_x, path_y = zip(*path)
                path_line.set_data(path_x, path_y)
        
        ani = animation.FuncAnimation(fig, update, frames=len(self.searched)+10, interval=50, repeat=False)
        ani.save("rrt_connect.gif", writer='pillow')
        plt.savefig("rrt_connect.png", bbox_inches='tight')
        plt.legend()
        plt.show()

File: RRT_Connect/CPP/src/rrt_connect.cpp
/**
 * @File: rrt_connect.cpp
 * @Brief: Description
 *
 * @Author: Benxiaogu
 * @Github: https://github.com/Benxiaogu
 * @CSDN: https://blog.csdn.net/weixin_51995147?type=blog
 *
 * @Date: 2024-11-13
 */

#include <iostream>
#include <vector>
#include <cmath>
#include <random>
#include <map>
#include <algorithm>
#include <opencv2/opencv.hpp>
#include "rrt_connect.hpp"

using namespace std;
using namespace cv;

// RRTConnect Constructor
RRTConnect::RRTConnect(pair<double,double> start, pair<double,double> goal, int width, int height, int board_size, int max_try, double max_dist, double goal_sample_rate, vector<Rect> boundary,vector<Rect> obs_rectangle, vector<Point3f> obs_circle)
    : start(start, start, 0), goal(goal, goal, 0), width(width), height(height), board_size(board_size), max_try(max_try),
      max_dist(max_dist), goal_sample_rate(goal_sample_rate), boundary(boundary), obs_rectangle(obs_rectangle), obs_circle(obs_circle) 
{
    inflation = 0.1;
}

void RRTConnect::run() 
{
    double cost=0;
    vector<pair<double,double>> path;
    auto result = plan(cost);
    path = result.first;
    searched = result.second;
    printf("Plan finished!\n");
    visualize(cost, path);
}

pair<vector<pair<double,double>>,vector<RRTConnect::Node>> RRTConnect::plan(double& cost) 
{
    // vector<pair<double,double>> path;
    map<pair<double,double>, Node> nodes_f = {{start.position, start}};
    map<pair<double,double>, Node> nodes_b = {{goal.position, goal}};

    for (int i = 0; i < max_try; ++i) {
        Node node_rand = get_random_node();
        Node node_near = get_nearest_neighbor(nodes_f, node_rand);
        Node node_new = get_new_node(node_rand, node_near);
        if (node_new.st) {
            nodes_f[node_new.position] = node_new;

            Node node_near_b = get_nearest_neighbor(nodes_b, node_new);
            Node node_new_b = get_new_node(node_new, node_near_b);

            if (node_new_b.st){
                nodes_b[node_new_b.position] = node_new_b;

                while (true)
                {
                    if (nodes_b.find(node_new.position) != nodes_b.end()){
                        auto [cost_path, path] = extractPath(node_new, nodes_f, nodes_b);
                        cost = cost_path;
                        vector<Node> expand = get_expand(nodes_f, nodes_b);
                        printf("Be in iteration %d.\n",i);

                        return {path, expand};
                    }

                    Node node_new_b2 = get_new_node(node_new, node_new_b);
                    if (node_new_b2.st){
                        nodes_b[node_new_b2.position] = node_new_b2;
                        node_new_b = node_new_b2;
                    }
                    else{
                        break;
                    }
                }
            }
        }

        if (nodes_b.size() < nodes_f.size()){
            swap(nodes_f,nodes_b);
        }
    }

    return {{},{}};
}

RRTConnect::Node RRTConnect::get_random_node() {
    random_device rd;
    mt19937 gen(rd());
    uniform_real_distribution<> dis_x(0, width), dis_y(0,height);
    if (dis_x(gen) > goal_sample_rate) {
        return Node(pair<double,double>(dis_x(gen), dis_y(gen)));
    } else {
        return Node(goal.position);
    }
}

RRTConnect::Node RRTConnect::get_nearest_neighbor(map<pair<double,double>, Node>& node_list, Node node) {
    double min_dist = numeric_limits<double>::max();
    Node nearest_node(node);
    for (auto& n : node_list) 
    {
        double dist = distance(n.second, node);
        if (dist < min_dist) {
            min_dist = dist;
            nearest_node = n.second;
        }
    }
    return nearest_node;
}

RRTConnect::Node RRTConnect::get_new_node(Node node_rand, Node node_near) {
    double dx = node_rand.position.first - node_near.position.first;
    double dy = node_rand.position.second - node_near.position.second;
    double dist = hypot(dx, dy);
    double theta = atan2(dy, dx);

    double d = min(max_dist, dist);
    pair<double,double> new_pos(node_near.position.first + d * cos(theta), node_near.position.second + d * sin(theta));

    Node node_new(new_pos, node_near.position, node_near.cost + d);
    if (isCollision(&node_new, &node_near)) {
        node_new.st = false;
    }
    else{node_new.st = true;}

    return node_new;
}

bool RRTConnect::isCollision(Node* node1, Node* node2) {
    if (!isInBoundary(node1)){
        return true;
    }
    if (isInObstacles(node1) || isInObstacles(node2)) {
        return true;
    }

    for (auto& rect : obs_rectangle) {
        if (isInterRect(node1, node2, rect)) {
            return true;
        }
    }

    for (auto& circle : obs_circle) {
        if (isInterCircle(node1, node2, circle)) {
            return true;
        }
    }

    return false;
}

bool RRTConnect::isInBoundary(Node* node)
{
    pair<double,double> node_position = node->position;
    int x1 = boundary[0].x+boundary[0].width;
    int y1 = boundary[1].y+boundary[1].height;
    int x2 = boundary[3].x;
    int y2 = boundary[2].y;
    // printf("x1: %d\n",x1);
    // printf("y1: %d\n",y1);
    // printf("x2: %d\n",x2);
    // printf("y2: %d\n",y2);

    if (node_position.first > x1 && node_position.first < x2 && node_position.second > y1 && node_position.second < y2)
    {
        return true;
    }
    return false;
}

bool RRTConnect::isInObstacles(Node* node) {
    pair<double,double> pos = node->position;
    Point2f point(pos.first, pos.second);

    for (auto& rect : obs_rectangle) {
        Rect rec = rect;
        rec.x -= inflation;
        rec.y -= inflation;
        rec.width += 2*inflation;
        rec.height += 2*inflation;
        if (rec.contains(point)) {
            return true;
        }
    }

    for (auto& circle : obs_circle) {
        double dist = hypot(pos.first - circle.x, pos.second - circle.y);
        if (dist <= circle.z+inflation) {
            return true;
        }
    }

    return false;
}

float RRTConnect::cross(const pair<double,double>& p1, const pair<double,double>& p2, const pair<double,double>& p3) {
    float x1 = p2.first - p1.first;
    float y1 = p2.second - p1.second;
    float x2 = p3.first - p1.first;
    float y2 = p3.second - p1.second;
    return x1 * y2 - x2 * y1;
}

bool RRTConnect::isInterRect(Node* node1, Node* node2, const cv::Rect& rect) {
    // Calculate the four vertices of the rectangle considering the inflation
    std::vector<pair<double,double>> vertex = {
        pair<double,double>(rect.x - inflation, rect.y - inflation),
        pair<double,double>(rect.x + rect.width + inflation, rect.y - inflation),
        pair<double,double>(rect.x + rect.width + inflation, rect.y + rect.height + inflation),
        pair<double,double>(rect.x - inflation, rect.y + rect.height + inflation)
    };

    pair<double,double> p1 = node1->position;
    pair<double,double> p2 = node2->position;

    // Check if any rectangle edge intersects the line segment from node1 to node2
    for (size_t i = 0; i < vertex.size(); ++i) {
        pair<double,double> v1 = vertex[i];
        pair<double,double> v2 = vertex[(i + 1) % vertex.size()];  // Ensuring cyclic order

        if (std::max(p1.first, p2.first) >= std::min(v1.first, v2.first) && std::min(p1.first, p2.first) <= std::max(v1.first, v2.first) &&
            std::max(p1.second, p2.second) >= std::min(v1.second, v2.second) && std::min(p1.second, p2.second) <= std::max(v1.second, v2.second)) {

            if (cross(v1, v2, p1) * cross(v1, v2, p2) <= 0 &&
                cross(p1, p2, v1) * cross(p1, p2, v2) <= 0) {
                return true;
            }
        }
    }

    return false;
}

bool RRTConnect::isInterCircle(Node* node1, Node* node2, Point3f circle) {
    // Calculate intersection with circle (similar to the Python implementation)
    double dx = node2->position.first - node1->position.first;
    double dy = node2->position.second - node1->position.second;

    if (dx * dx + dy * dy == 0){
        return false;
    }

    double t = ((circle.x - node1->position.first) * dx + (circle.y - node1->position.second) * dy) / (dx * dx + dy * dy);

    if (0 <= t && t <= 1) {
        double closest_x = node1->position.first + t * dx;
        double closest_y = node1->position.second + t * dy;
        double distance = hypot(circle.x - closest_x, circle.y - closest_y);

        return distance <= circle.z+inflation;
    }

    return false;
}

double RRTConnect::distance(Node node1, Node node2) 
{
    return hypot(node2.position.first - node1.position.first, node2.position.second - node1.position.second);
}

pair<double,vector<pair<double,double>>> RRTConnect::extractPath(Node node_middle, 
                                                    map<pair<double,double>, Node> nodes_f, 
                                                    map<pair<double,double>, Node> nodes_b) 
{
    if (nodes_b.find(start.position) != nodes_b.end()) {
        std::swap(nodes_f, nodes_b);
    }

    Node node = nodes_f[node_middle.position];

    vector<pair<double,double>> path_f = {node.position};
    double cost = node.cost;
    while (node.position != start.position) 
    {
        node = nodes_f[node.parent];
        path_f.push_back(node.position);
    }
    reverse(path_f.begin(), path_f.end());

    node = nodes_b[node_middle.position];
    vector<pair<double,double>> path_b;
    cost += node.cost;
    while (node.position != goal.position)
    {
        node = nodes_b[node.parent];
        path_b.push_back(node.position);
    }

    path_f.insert(path_f.end(),path_b.begin(),path_b.end());

    return {cost, path_f};
}

vector<RRTConnect::Node> RRTConnect::get_expand(map<pair<double,double>,Node>& nodes_f, map<pair<double,double>,Node>& nodes_b)
{
    vector<Node> expand;
    size_t tree_size = max(nodes_f.size(),nodes_b.size());

    auto it_forward = nodes_f.begin();
    auto it_back = nodes_b.end();
    for (size_t i=0; i<tree_size; i++)
    {
        if (it_forward != nodes_f.end())
        {
            expand.push_back(it_forward->second);
            ++it_forward;
        }
        if (it_back != nodes_b.begin())
        {
            expand.push_back(it_back->second);
            --it_back;
        }
    }

    return expand;
}

void RRTConnect::visualize(double cost, vector<pair<double,double>>& path) {
    int cell_size = 20;

    VideoWriter video("RRTConnect_animation.mp4", VideoWriter::fourcc('m', 'p', '4', 'v'), 100, Size(width*cell_size, height*cell_size));

    Mat img = Mat::zeros(height*cell_size, width*cell_size, CV_8UC3);
    img.setTo(Scalar(255, 255, 255));

    // Title and Cost
    std::string title = "RRTConnect";
    putText(img, title, Point(550, 20), FONT_HERSHEY_SIMPLEX, 0.8, Scalar(0, 0, 0), 2);
    std::string cost_text = "Cost: " + std::to_string(cost);
    putText(img, cost_text, Point(500, 50), FONT_HERSHEY_SIMPLEX, 0.8, Scalar(0, 0, 0), 2);

    // Draw boundary
    for (auto& rect : boundary) {
        rect.x *= cell_size;
        rect.y *= cell_size;
        rect.height *= cell_size;
        rect.width *= cell_size;

        rectangle(img, rect, Scalar(0, 0, 0), FILLED);
    }
    // Draw obstacles
    for (auto& rect : obs_rectangle) {
        rect.x *= cell_size;
        rect.y *= cell_size;
        rect.height *= cell_size;
        rect.width *= cell_size;

        rectangle(img, rect, Scalar(0, 0, 0), FILLED);
    }
    for (auto& circ : obs_circle) {
        circle(img, Point(circ.x*cell_size, circ.y*cell_size), circ.z*cell_size, Scalar(0, 0, 0), FILLED);
    }

    // Draw start and goal
    circle(img, Point2f(start.position.first*cell_size,start.position.second*cell_size), 5, Scalar(0, 255, 0), FILLED);
    circle(img, Point2f(goal.position.first*cell_size,goal.position.second*cell_size), 5, Scalar(255, 0, 0), FILLED);

    // Draw visited areas
    for (size_t i=1;i<searched.size();++i) 
    {
        Node node = searched[i];
        line(img, Point2f(node.position.first*cell_size,node.position.second*cell_size), Point2f(node.parent.first*cell_size,node.parent.second*cell_size), Scalar(0, 100, 0), 2);
        
        video.write(img);
    }

    // Draw the path
    for (size_t i = 0; i < path.size()-1; ++i) {
        line(img, Point2f(path[i].first*cell_size,path[i].second*cell_size), Point2f(path[i + 1].first*cell_size,path[i+1].second*cell_size), Scalar(0, 0, 255), 2);
        video.write(img);
    }

    video.release();
    // Show the image
    imwrite("RRTConnect.png", img);
    imshow("RRTConnect Path", img);
    waitKey(0);
    destroyAllWindows();
    }